### ISSUE TITLE
[BUG] Creating/Syncing Multiple Forever-Tasks (e.g. Daily Tasks) Causes Unusable State

### ISSUE THREAD

**Issue by jackrgm** [TechbeeAT/jtxBoard#1071]
I assume it's a valid use of CalDAV to create routine tasks - e.g. practice programming, everyday, forever - but for me it seems that setting up multiple of these daily or weekly tasks lead to jtx Board becoming unusable (slow, not synching correctly).

Perhaps it's to do with how recurring tasks are made? Looking at the series for one recurring task, I can see that it's set up with 588 instances.

For now I'll have to find some other way to achieve this, but it would be great to be able to do it via CalDAV and jtx Board.

**Device and Version**
 * Device: Fairphone 4
 * Android version: Android 13 (LineageOS)
 * jtx Board version: 2.06.04.ose (206040012)
 * Downloaded from: F-Droid

**Other Information**

I use Gnome Evolution to sync my tasks, calendar, notes (memos), and journals (memos) on desktop.

**Comment by jackrgm** [comment]
My apologies, I hadn't thought to stop the jtx Board app and clear its cache before posting this. Nevermind. :)

After stopping the app and clearing its cache the app responsiveness and syncing seems to be okay now. Takes a second to load up the task list (recurring tasks leads to lots of entries I guess), but it's generally usable from what I see now.

**Comment by jackrgm** [comment]
Unfortunately, after trying to use jtx Board some more, it seems the issue remains.

But I don't get problems with syncing, rather it's that task details aren't loading when I tap on them, it just shows a blank page, and overall the app goes clunky.

I think it's related to the fact that having multiple daily/weekly repeat tasks (each adding about 500 entries) result in 4000+ tasks overall, according to the app. 'Limit recurring' is on.

Gnome Evolution doesn't seem to have this issue, though it only shows one instance at all, then creates the next instance when the current one is marked completed.

**Comment by patrickunterwegs** [comment]
Hey @jackrgm , 
thank you for the bug report. 
So I understand the the problem can be broken down to this: 

> But I don't get problems with syncing, rather it's that task details aren't loading when I tap on them, it just shows a blank page, and overall the app goes clunky.

No issues in the list, right? Only when you open the details screen?
Do you also get any messages that the app is not responding or that the opened entry doesn't exist anymore or does the screen just stay empty?

I'm wondering what could really cause this issue. The data is loaded asynchronously, so basically there is no reason why data doesn't show up...

**Comment by jackrgm** [comment]
Apologies for the late reply, I underestimated the Christmas turkey.

I took some time to fiddle about and do some amateur QA testing. I wrote down my steps and findings below to help this issue be more repeatable and help explain what I'm experiencing.

Hopefully this is insightful.

## Testing Steps

**Note:** 'limit recurring' was enabled, all task fields default except where noted.

1. Create CalDAV collection 'tmp_jtx_recurringbug'. Disable all other collections in DAVx5.
2. Create task 'SINGLE'. jtx Board reports 1/1 tasks in list.
	a. Leave app after creating task. Wait for app to unload or force stop the app.
	b. Return to app. Open task to view details, then return to task list. Repeat at least four times; loading of details page generally feels responsive.
	c. Delete task 'SINGLE'. jtx Board reports 0/0 tasks in list.
5. Create 'RECUR_1' - start today at 9am. Repeat every `1` `day(s)` `ends never`. jtx reports 1/366 total entries in the task list.
	a. Leave and return to app as in step 2.
	b. Open task to view details, then return to task list. Repeat at least four times; loading of details page includes a quick pause (82ms avg. on the stopwatch?) but otherwise feels responsive.
3. Create 'RECUR_2' - same task settings as RECUR_1. jtx Board now reports 2/732 total entries (366 new recurring entries added).
	a. As in previous steps, leave and return to app, open task(s) to view details. About the same level of responsiveness.
4. Create 'RECUR_3' - same task settings as RECUR_1. jtx Board now reports 3/1252 total entries (not sure why 520 recurring entries were added this time).
	a. As in previous steps, leave and return to app, open task(s) to view details. Task list starts empty with a noticable pause before entries quickly appear. View details page is perhaps the same, perhaps slower, hard to say.
5. Create RECUR_4 - same as RECUR_1. jtx Board now reports 4/1618 total entries (it created 366 recurring entries this time).
	a. As in previous steps, leave and return, etc. - Task list now starts empty with a considerable pause time (about twice as long as last time) before entries appear. Viewing task details has a varying rate of responsiveness each time, but generally a noticeable sense of slowness.
6. Create RECUR_5 - same as RECUR_1. jtx Board now reports 5/1984 total entries (366 new). Experienced general sluggish and slow performance while creating the task.
	a. As in previous steps, leave, return, etc. - task list is even slower than last time in loading entries in the list (I suspect this will grow for each daily recurring task added). Viewing task details responsiveness varies, but generally feels sluggish and slow to load.

## Other Observations Since Creating 'RECUR_5'

* Upon creating a new task, the new task window now takes a significant time to load
* I tried creating a note and it took a considerable amount of time to load the new entry page.
* The app generally feels slow all around at this point.
* Added category 'TESTCAT' to 'RECUR_5'. At this point, after saving my changes, then pressing the back button, the app ignored the first button press, then took several seconds to go back after pressing the back button again.
* Filtering for 'TESTCAT' showed results quick. Resetting filters took a quick pause, but was generally responsive.
* Created task 'SYNCTEST' on Gnome Evolution from Linux desktop; all defaults, non-recurring. This synchronised fine with jtx Board with no noticeable issues.
* Created task 'RECUR_FROM_EVO' on Gnome Evolution from Linux desktop; start at 9am today, repeat every day forever. jtx Board took at least 25 seconds to complete a synchronise action.
* Calling a synchronise action after this (no changes made) is generally quick, no issues noticed there.
* Sometimes viewing the details of a task at this point could take a good second or two for the page to go from blank (nothing but the search bar) to task details appearing.
* No warnings, error messages, or anything like that (including 'app not responding' notices) appeared during this process. It's mostly a performance thing, by the looks of it anyway.
* Deleting/disabling the test collection (tmp_jtx_recurringbug) and re-enabling my personal collections (which currently don't include any recurring tasks) returns jtx Board to a responsive, good performance state. jtx Board currently reports 565/627 task entries with my personal collections active.

**Comment by patrickunterwegs** [comment]
Hey @jackrgm , sorry that it took me a while to come back to this. I was very busy and it took me a while to figure out what could be the cause. I think I found it now, also thanks to your detailed checks. 
I will upload a new version soon that hopefully fixes the issue also for you!

**Comment by patrickunterwegs** [comment]
So I have added the change in 2.07.02-alpha10, I expect to release 2.07.02 in the upcoming two weeks. I'd appreciate
your feedback once it's also available for you!

**Comment by jackrgm** [comment]
Sure thing, I look forward to checking it out. Glad to be of help and big thanks for all the awesome work you're doing on jtxBoard and for expanding the CalDAV world. 🙂

**Comment by sssammm88** [comment]
Thanks for this great app @patrickunterwegs, which I discovered & adopted recently.

I also have quite a bunch of recurring & non recurring tasks myself in the app (about 5800 tasks are shown in the app) and I experience the same sluggishness & issues as described in this ticket. I strongly suspect this is related to the recurrence feature.

In order to see if it was the case, I also did some testing myself. 

# Configuation

Device: Google Pixel pro 8
OS: GrapheneOS (Android 15)
Version: 2.09.02.ose (209020004)
Downloaded from: F-Droid
Synchronization: DAVx5 (4.4.4-ose) on Nextcloud Server

# Test protocol

I replicated the protocol of @jackrgm 

> ## Testing Steps
> 
> **Note:** 'limit recurring' was enabled, all task fields default except where noted.
> 
>     1. Create CalDAV collection 'tmp_jtx_recurringbug'. Disable all other collections in DAVx5.
> 
>     2. Create task 'SINGLE'. jtx Board reports 1/1 tasks in list.
>        a. Leave app after creating task. Wait for app to unload or force stop the app.
>        b. Return to app. Open task to view details, then return to task list. Repeat at least four times; loading of details page generally feels responsive.
>        c. Delete task 'SINGLE'. jtx Board reports 0/0 tasks in list.
> 
>     3. Create 'RECUR_1' - start today at 9am. Repeat every `1` `day(s)` `ends never`. jtx reports 1/366 total entries in the task list.
>        a. Leave and return to app as in step 2.
>        b. Open task to view details, then return to task list. Repeat at least four times; loading of details page includes a quick pause (82ms avg. on the stopwatch?) but otherwise feels responsive.
> 
>     4. Create 'RECUR_2' - same task settings as RECUR_1. jtx Board now reports 2/732 total entries (366 new recurring entries added).
>        a. As in previous steps, leave and return to app, open task(s) to view details. About the same level of responsiveness.
> 
>     5. Create 'RECUR_3' - same task settings as RECUR_1. jtx Board now reports 3/1252 total entries (not sure why 520 recurring entries were added this time).
>        a. As in previous steps, leave and return to app, open task(s) to view details. Task list starts empty with a noticable pause before entries quickly appear. View details page is perhaps the same, perhaps slower, hard to say.
> 
>     6. Create RECUR_4 - same as RECUR_1. jtx Board now reports 4/1618 total entries (it created 366 recurring entries this time).
>        a. As in previous steps, leave and return, etc. - Task list now starts empty with a considerable pause time (about twice as long as last time) before entries appear. Viewing task details has a varying rate of responsiveness each time, but generally a noticeable sense of slowness.
> 
>     7. Create RECUR_5 - same as RECUR_1. jtx Board now reports 5/1984 total entries (366 new). Experienced general sluggish and slow performance while creating the task.
>        a. As in previous steps, leave, return, etc. - task list is even slower than last time in loading entries in the list (I suspect this will grow for each daily recurring task added). Viewing task details responsiveness varies, but generally feels sluggish and slow to load.

# Results

I got essentially the same results as in the post above, with some small differences:

* The "correct" amount of new tasks was created every time (366)
* I didn't experience sluggishness when going to the description & back to the main menu during the test, but I do experience it in my normal "big" tasks list
* Every addition of a daily infinite recurring time increased the sync delay with the server of about 3 seconds (9 seconds with 3 tasks, 12 with 4, 15 with 5)
* Starting from 5 recurring tasks, I had to wait about 1 second between pressing the checkbox icon and having the task being marked as complete. In my usual big list of tasks, the delay is about 10-15 seconds. This is true even in the absence of any task modification
* In the test mode, if I delete 4 out of 5 recurring tasks, the app is fully reactive again.
* After starting the resyncing the original list of tasks, I get the following error: `Local storage error : Couldn't apply batch operation`. Retriggering the sync increases the number of displayed list of tasks of about 500

To summarize, the app is functional, but very slow in my usecase.

Let me know if I can help further here (eg. via testing) and thanks again for the work put in this app. I really appreciate when someone puts value in correct standard implementation (which is far from being the norm with task management app IMO).

ps. I'm quite new to contributing to open source projects. My apologies if I overlooked a code of contribution or something similar.
